source/tutorials/callpackage.md

```diff
@@ -0,0 +1,283 @@
+Change the `default.nix` to produce an attribute set of derivations, with the attribute `hello` containing the original derivation:
+
+```{code-block} nix
+:caption: default.nix
+let
+  pkgs = import <nixpkgs> { };
+in
+{
+  hello = pkgs.callPackage ./hello.nix { };
+}
+```
+
+When building the attribute `hello`, by accessing it with the [`-A` / `--attr` option](https://nix.dev/manual/nix/2.19/command-ref/nix-build#opt-attr), the result will be the same as before:
+
+```shell-session
+$ nix-build -A hello
+$ ./result/bin/hello
+Hello, world!
+```
+
+Also change `hello.nix` to add an additional parameter `audience` with default value `"world"`:
+
+```{code-block} nix
+:caption: hello.nix
+{
+  writeShellScriptBin,
+  audience ? "world",
+}:
+writeShellScriptBin "hello" ''
+  echo "Hello, ${audience}!"
+''
+```
+
+This also does not change the result.
+
+Things get more interesting when changing `default.nix` to make use of this new argument.
+Pass the parameter `audience` in the second argument to `callPackage`:
+
+```{code-block} diff
+:caption: default.nix
+ let
+   pkgs = import <nixpkgs> { };
+ in
+ {
+   hello = pkgs.callPackage ./hello.nix { };
++  hello-people = pkgs.callPackage ./hello.nix { audience = "people"; };
+ }
+```
+
+This attribute is passed on to the argument of the function defined in `hello.nix`:
+The same syntax can also be used to explicitly set the automatically discovered arguments, such as `writeShellScriptBin`, but that doesn't make sense here.
+
+Try it out:
+
+```shell-session
+$ nix-build -A hello-people
+$ ./result/bin/hello
+Hello, people!
+```
+
+This pattern is used widely in Nixpkgs:
+For example, functions which represent Go programs often have a parameter `buildGoModule`.
+It is common to find expressions like `callPackage ./go-program.nix { buildGoModule = buildGo116Module; }` to enforce a certain Go compiler version.
+Nixpkgs is therefore not simply a huge library of pre-configured packages, but a collection of functions – package *recipes* – for customising packages and even entire ecosystems (for example "All Python packages using my custom interpreter") on the fly without duplicating code.
+
+# Overrides
+
+`callPackage` adds more convenience by adding an attribute to the derivation it returns: the `override` function.
+This means that, as a consequence of parameterising builds with `callPackage`, you can also change the value of these parameters _after the fact_ using the returned derivation's `override` function.
+
+Add a third attribute `hello-folks` to `default.nix` and set it to `hello.override` called with a new value for `audience`:
+
+```{code-block} diff
+:caption: default.nix
+ let
+   pkgs = import <nixpkgs> { };
+ in
+-{
++rec {
+   hello = pkgs.callPackage ./hello.nix { };
+   hello-people = pkgs.callPackage ./hello.nix { audience = "people"; };
++  hello-folks = hello.override { audience = "folks"; };
+ }
+```
+
+:::{note}
+The resulting attribute set is now recursive (by the keyword `rec`).
```

Comment:
Would be good to link to https://nix.dev/guides/best-practices#recursive-attribute-set-rec here

Reply:
I've thought about it, but would rather skip it here:
- It's besides the point. Adding more comments will distract from the actual learning objective. The note is merely to remind novice readers what `rec` is about so they can fully grasp the example.
- Referring to a piece that says "`rec` bad" from a place where we very deliberately use `rec` for clarity and small diff seems weird and counterproductive to building trust in the information we provide.
- Thinking about "`rec` bad" made me doubt this should be here on nix.dev to begin with. It seems it's more of a Nixpkgs convention, where it makes sense if you have huge files and the need to separate interface from implementation. But generally, using `rec` judiciously can greatly increase readability in my opinion.